packages/dds/sequence/src/sharedStringWithSequentialId.ts

```diff
@@ -0,0 +1,89 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IMergeTreeOp, Marker,
+    MergeTreeDeltaType, reservedMarkerIdKey } from "@fluidframework/merge-tree";
+import { ISequencedDocumentMessage, MessageType } from "@fluidframework/protocol-definitions";
+import { parseHandles } from "@fluidframework/shared-object-base";
+import { createIdAfterMin, createIdBeforeMax } from "./generateSequentialId";
+import { SharedString } from "./sharedString";
+
+export function sharedStringWithSequentialIdMixin(Base: typeof SharedString = SharedString): typeof SharedString {
+    return class SequenceWithSequentialId extends Base {
+        protected processMergeTreeMsg(
+            rawMessage: ISequencedDocumentMessage): void {
+                const message = parseHandles(
+                    rawMessage,
+                    this.runtime.IFluidSerializer);
+                this.applySequentialId(message);
+                super.processMergeTreeMsg(rawMessage);
+        }
+
+        private applySequentialId(msg: ISequencedDocumentMessage) {
+            if (msg.type !== MessageType.Operation) {
+                return;
+            }
+
+            const op = msg.contents as IMergeTreeOp;
+            if (op.type === MergeTreeDeltaType.INSERT && op.seg !== undefined
+                && typeof op.seg === "object") {
+                // this is a marker
+                if (op.pos1 === undefined) {
+                    return;
+                }
+
+                const insertSegment = this.client.getContainingSegment(op.pos1);
+                if (!Marker.is(insertSegment.segment)) {
+                    return;
+                }
+                const newMarker = insertSegment.segment;
+                let newMarkerProps = newMarker.properties ?? {};
+                const newMarkerCp = this.getPosition(newMarker);
+                const beforeMarkers: Marker[] = [];
+                const afterMarkers: Marker[] = [];
+
+                this.walkSegments((segment) => {
```

Comment:
i'm really worried about performance here, and this an the below will walk every segment in the merge tree on every insert which i don't this is sustainable.

Reply:
Can we solve it by some kind of hash map?

Reply:
i think we'll need some kind of index. take a look at sorted segment set in merge tree

---
In reply to: [546021358](https://github.com/microsoft/FluidFramework/pull/4641#discussion_r546021358) [](ancestors = 546021358)

Reply:
[https://github.com/microsoft/FluidFramework/blob/e5a7147aad10d055a12086d37e64080317131bb9/packages/dds/merge-tree/src/sortedSegmentSet.ts](https://github.com/microsoft/FluidFramework/blob/e5a7147aad10d055a12086d37e64080317131bb9/packages/dds/merge-tree/src/sortedSegmentSet.ts)


---
In reply to: [546048239](https://github.com/microsoft/FluidFramework/pull/4641#discussion_r546048239) [](ancestors = 546048239,546021358)

Reply:
Made changes to use sortedSegmentSet. After chatting with Tony, I'm making this set lazy to avoid being in the critical path and populating it the first time it is going to be used.

From conversation with Tony this is not the ideal solution but we can later revisit if there is a way to persist this set